### ISSUE TITLE
Unable to resolve @angula/core module when there is a symlinked package

### ISSUE THREAD

**Issue by Fatme** [angular/angular-cli#15604]
# 🐞 Bug report

### Command (mark with an `x`)

```
- [ ] new
- [x] build
- [x] serve
- [x] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [x] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Is this a regression?

No

### Description

In case we have an application with symlinked typescript package (a typescript project that is required locally from another folder), `ERROR in Could not resolve module @angular/core` error is thrown. This happens when the symlinked package is added as first element in files property in `tsconfig.app.json` file [as shown here in demo app](https://github.com/Fatme/testApp/blob/master/tsconfig.app.json).

It seems that `@angular/core` module is required in source code [here](https://github.com/angular/angular/blob/ec4381dd401f03bded652665b047b6b90f2b425f/packages/compiler-cli/src/transformers/program.ts#L562) in order to resolve all decorated elements from source file. It seems that `@angular/core` is searched for as a dependency relatively to the [first included source file](https://github.com/angular/angular/blob/3166cffd28e12d882bef3e55472e0a50f4cfc6f5/packages/compiler-cli/src/transformers/compiler_host.ts#L192). Source files are gettered based on `tsconfig.json` file and mainly on `files`, `include` and `exclude` properties. However, in our particular case when there is a symlinked package, it matters if `@angular/core` will be searched for relatively to the symlinked package or relatively to the application as `@angular/core` is dependency in the application but it is not dependency of the symlinked package.

The comment [here](https://github.com/angular/angular/blob/3166cffd28e12d882bef3e55472e0a50f4cfc6f5/packages/compiler-cli/src/transformers/compiler_host.ts#L191) says that any containing file gives the same result for absolute imports. As we can see, there are situations when the order of the imported files affect the final result:
* when file outside of the application is first element in files property in `tsconfig.app.json` file -> it is not possible to resolve `@angular/core`
* when file from application is first element in files property in `tsconfig.app.json` file -> `@angular/core` is resolved correctly

It seems that getting the first source file as containing folder is not the correct decision, maybe all files should be checked until `@angular/core` is found.

## 🔬 Minimal Reproduction

1. git clone https://github.com/Fatme/symlinkedPlugin
2. git clone https://github.com/Fatme/testApp
3. cd symlinkedPlugin
4. `npm i`
5. cd ../testApp
6. `npm i`
7. `ng serve`

> NOTE: If you change the order of files [here](https://github.com/Fatme/testApp/blob/master/tsconfig.app.json#L8) and `main.ts` is the first file, the issue will not reproduce.

## 🔥 Exception or Error
<pre><code>
ERROR in Could not resolve module @angular/core

</code></pre>

## 🌍 Your Environment
<pre><code>
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 8.3.4
Node: 10.13.0
OS: darwin x64
Angular: 8.2.6
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.803.4
@angular-devkit/build-angular     0.803.4
@angular-devkit/build-optimizer   0.803.4
@angular-devkit/build-webpack     0.803.4
@angular-devkit/core              8.3.4
@angular-devkit/schematics        8.3.4
@angular/cli                      8.3.4
@ngtools/webpack                  8.3.4
@schematics/angular               8.3.4
@schematics/update                0.803.4
rxjs                              6.4.0
typescript                        3.5.3
webpack                           4.39.2

</code></pre>

**Comment by alan-agius4** [comment]
Hi there, using typescript path mapping for dependencies should address your issue. See https://angular.io/guide/creating-libraries#use-typescript-path-mapping-for-peer-dependencies

Also, it's important to mention that it is discouraged to compile a separate module/library using the applications compilation. This is because libraries get compiled using a different pipeline and may result in inconsistency between a standard library build and an application build.

**Comment by Fatme** [comment]
Hey @alan-agius4,

The downgrade of using typescript path mapping for dependencies is that the library should use exactly the same version of `@angular/*` packages as the application i.e when the application is updated to the newer version of `@angular/*`, the library should be updated as well. I'm not sure if this is the intended and recommended behavior for angular libraries.

However, in my particular case your suggestion will work as my symlinked package doesn't rely on angular at all i.e it is not an angular library.

Regarding to your concern that libraries shouldn't be compiled using the compilation pipeline of the application - Putting code into a separate non-Angular library shouldn't be more complex than simply putting everything in the application. It shouldn't result in a such unexpected behavior that is relying on the order of the included source files.

**Comment by alan-agius4** [comment]
> The downgrade of using typescript path mapping for dependencies is that the library should use exactly the same version of @angular/* packages as the application i.e when the application is updated to the newer version of @angular/*, the library should be updated as well. I'm not sure if this is the intended and recommended behavior for angular libraries.

Yeah, this is indented and recommended, since a library should use `peerDepedencies` instead of `dependencies` to avoid having duplicate dependencies in the final bundle which will result in an increase bundle size.


> However, in my particular case your suggestion will work as my symlinked package doesn't rely on angular at all i.e it is not an angular library.

 That is true, and this seem like a bug in the compiler cli. 

> Regarding to your concern that libraries shouldn't be compiled using the compilation pipeline of the application - Putting code into a separate non-Angular library shouldn't be more complex than simply putting everything in the application. It shouldn't result in a such unexpected behavior that is relying on the order of the included source files.

In your case indeed it is simple, but if you have a component for instance styles are compiled differently and there are certain restrictions by design.
